### ISSUE TITLE
Knowledge Sharing - Redis/Envoy Ratio

### ISSUE THREAD

**Issue by yairgott** [envoyproxy/envoy#25332]
Hey Envoy Redis experts,

We're evaluating Envoy as a Redis Proxy. I would appreciate it if you can share some data about others' experience with regards to what is the expected ratio of the overall cluster CPU between Redis and Envoy instances in real production environments.

I can share that running some benchmarks in the lab, simple get:set commands with the ratio of 3:1 with a set size of 1kb, we see a ratio of about 2:1, Envoy:Redis. 

Logically, it makes sense to me that Envoy consumes more CPU as it's doing more work per command, dealing with network on both sides, while Redis handles network on just one side and performs o(1) ,set/get, operations. In this case, handling the network, which involves frequent context switching between kernel and user space, seems to me much more expensive than executing the command itself. Does that sound right?

Also, I would appreciate it if you can share some data about Envoy as Redis proxy tuning to get the most out of it.  

Thanks in advance,
Yair

**Comment by wbpcode** [comment]
cc @weisisea

**Comment by weisisea** [comment]
Hi Yair, I think your understanding is correct. Assuming that the number of Envoy instances is the same as Redis instances, an Envoy instance processes double the network traffic as an Redis instance does.

In terms of tuning Envoy for performance, we have found adjusting the two parameters could improve the throughput a little:
1. Increasing `max_buffer_size_before_flush` ([link](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/redis_proxy/v3/redis_proxy.proto#extensions-filters-network-redis-proxy-v3-redisproxy-connpoolsettings)) - Potentially reduces number of writev syscalls.
2. Decreasing `per_connection_buffer_limit_bytes` ([link](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener.proto#config-listener-v3-listener)) - Potentially reduces amount of connection buffer allocation.

Some other areas are also worth exploring:
* Use newer Linux kernel
* Tune Linux kernel parameters such as TCP buffer sizes
* Disable conntrack (we found that conntrack takes ~2% CPU usage in flame chart)
* Use larger network MTU

We are also looking at some parts of Redis filter and see whether we could reduce their CPU cycles, e.g., metric aggregation. We'll post the changes once they are evaluated and developed.

Hope it helps.
